### PR TITLE
Support references in i256 arithmetic ops

### LINKED ISSUE #4694: Support references in i256 arithmetic ops
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Got `cannot add &i256 to &i256` while working with i256 value.

**Describe the solution you'd like**

Make i256 arithmetic operations support references.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #4694.

# Rationale for this change
 

Avoid errors like `cannot add &i256 to &i256` while working with i256 value.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-buffer/src/bigint/mod.rs [hunk 1]
```diff
@@ -659,6 +659,48 @@ macro_rules! derive_op {
                 self.$wrapping(rhs)
             }
         }
+
+        impl<'a> std::ops::$t<i256> for &'a i256 {
+            type Output = i256;
+
+            #[cfg(debug_assertions)]
```
Comment (tustvold) [inline arrow-buffer/src/bigint/mod.rs:666]: Could we just dereference the appropriate side as opposed to duplicating the debug_assertions logic?

Comment (viirya) [reply arrow-buffer/src/bigint/mod.rs:666]: I think we don't have dereference implementation for i256? Or I misunderstand your comment?

```
error[E0614]: type `bigint::i256` cannot be dereferenced
```

Comment (tustvold) [reply arrow-buffer/src/bigint/mod.rs:666]: So something like
```
impl<'a> std::ops::$t<i256> for &'a i256 {
        type Output = i256;
        fn $op(self, rhs: i256) -> Self::Output {
                (*self).$op(rhs)
        }
}
```

Comment (viirya) [reply arrow-buffer/src/bigint/mod.rs]: Oops, I see. Thanks.

**Comment by viirya** [comment]
Thanks @tustvold

**Pull request merged**